core/arch/arm/kernel/sub.mk

```diff
@@ -70,6 +70,15 @@ srcs-y += link_dummies_paged.c
 srcs-y += link_dummies_init.c
 
 asm-defines-y += asm-defines.c
+# Reflect the following dependencies:
+# asm-defines.c includes <kernel/thread.h>
+#   <kernel/thread.h> includes <asm.h>
+#     <asm.h> includes <generated/arm32_sysreg.h>
+#                  and <generated/arm32_gicv3_sysreg.h> (optional)
+asm-defines-asm-defines.c-deps += $(out-dir)/core/include/generated/arm32_sysreg.h
```

Comment:
This adds a direct dependency for asm-defines.c, but isn't this just the tip of an iceberg? There's bound to be plenty of other .c files depending on these .h files. Don't they cause the same kind of problems?

Reply:
Very good question :) The answer is no, because all "normal" C files have a dependency on `FORCE-GENSRC$(sm)` which includes the generation of both the sysreg files and the ASM defines, as shown below:

```
$ rm -rf out; make --debug=b out/arm-plat-vexpress/core/arch/arm/kernel/boot.o
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'out/arm-plat-vexpress/core/arch/arm/kernel/boot.o' does not exist.
   File 'FORCE-GENSRCcore' does not exist.
     File 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.h' does not exist.
    Must remake target 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.h'.
  GEN     out/arm-plat-vexpress/core/include/generated/arm32_sysreg.h
    Successfully remade target file 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.h'.
     File 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.S' does not exist.
    Must remake target 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.S'.
  GEN     out/arm-plat-vexpress/core/include/generated/arm32_sysreg.S
    Successfully remade target file 'out/arm-plat-vexpress/core/include/generated/arm32_sysreg.S'.
     File 'out/arm-plat-vexpress/core/include/generated/asm-defines.h' does not exist.
       File 'out/arm-plat-vexpress/core/include/generated/.asm-defines.s' does not exist.
         File 'out/arm-plat-vexpress/include/generated/conf.h' does not exist.
           File 'FORCE' does not exist.
          Must remake target 'FORCE'.
          Successfully remade target file 'FORCE'.
           File 'out/arm-plat-vexpress/conf.mk' does not exist.
          Must remake target 'out/arm-plat-vexpress/conf.mk'.
  CHK     out/arm-plat-vexpress/conf.mk
  UPD     out/arm-plat-vexpress/conf.mk
          Successfully remade target file 'out/arm-plat-vexpress/conf.mk'.
        Must remake target 'out/arm-plat-vexpress/include/generated/conf.h'.
  CHK     out/arm-plat-vexpress/include/generated/conf.h
  UPD     out/arm-plat-vexpress/include/generated/conf.h
        Successfully remade target file 'out/arm-plat-vexpress/include/generated/conf.h'.
      Must remake target 'out/arm-plat-vexpress/core/include/generated/.asm-defines.s'.
  CC      out/arm-plat-vexpress/core/include/generated/.asm-defines.s
      Successfully remade target file 'out/arm-plat-vexpress/core/include/generated/.asm-defines.s'.
    Must remake target 'out/arm-plat-vexpress/core/include/generated/asm-defines.h'.
  CHK     out/arm-plat-vexpress/core/include/generated/asm-defines.h
  UPD     out/arm-plat-vexpress/core/include/generated/asm-defines.h
    Successfully remade target file 'out/arm-plat-vexpress/core/include/generated/asm-defines.h'.
  Must remake target 'FORCE-GENSRCcore'.
  Successfully remade target file 'FORCE-GENSRCcore'.
Must remake target 'out/arm-plat-vexpress/core/arch/arm/kernel/boot.o'.
  CC      out/arm-plat-vexpress/core/arch/arm/kernel/boot.o
Successfully remade target file 'out/arm-plat-vexpress/core/arch/arm/kernel/boot.o'.
```

So the problem is limited to dependencies within this "generated sources" target, and therefore this ad-hoc fix is correct I believe.

I will add some text to the commmit description to clarify: "Other core `.c` files depending on these generated sysreg headers are not affected because their `.o` explicitly depend on `FORCE-GENSRCcore`.".

Reply:
Aha, thanks for looking into this.